File: active_set.py
import numpy as np
from scipy import sparse
from sksparse.cholmod import cholesky

class Active_set:

    def __init__(self, H, f, A, b):

        # Primal Formulation
        self.m = A.shape[0] #number of constraints
        self.H = H
        self.f = f
        self.A = A
        self.b = b
        self.max_iter = 8
        #tolerance
        self.eps_p = 1e-6
        self.eps_d = 1e-12
        self.eps_z = 1e-11 #zero

        self.lamda = None
        
        # Dual Formulation
        self.R = np.transpose(np.linalg.cholesky(self.H))
        self.Rinv = np.linalg.inv(self.R)
        # print(self.R)

        if not np.allclose(self.R, np.triu(self.R)):# check if upper triangular
            print("Not Upper!")

        self.M = None; self.v = None; self.d = None;

        #State Variables:
        self.lamda = None
        self.W = None
        self.W_hat = None

    def form_dual_objective(self):

        self.M = self.A.dot(self.Rinv)
        Rinv_T = self.Rinv.T
        self.v = Rinv_T.dot(self.f)
        self.d = self.b + self.M.dot(self.v)

        # print(self.M)
        # print(self.v)
        # print(self.d)

def solve(Prob):
    
    ## Working Set. Will keep increasing with iterations ##
    Prob.W = np.array([0])
    Prob.W_hat = np.setdiff1d(np.arange(Prob.m),Prob.W)
    Prob.lamda = np.zeros((Prob.m,1))
    Prob.lamda[0] = 3.0
    # print(Prob.W_hat)
    # print(Prob.W)
    # print(Prob.lamda)
    k = 0
    lamda_k = None
    lamda_k_W = None

    while k<Prob.max_iter:
        print(f'Iteration: {k+1}')
        M_k = Prob.M[Prob.W]
        # print("M_k: ", M_k)
        d_k = Prob.d[Prob.W]
        # print("d_k: ", d_k)

        M_k_hat = Prob.M[Prob.W_hat]
        d_k_hat = Prob.d[Prob.W_hat]
        mu_k = np.zeros((Prob.m, 1))
        ############ Process ###########
        '''
        MMT = (M_k)(M_k.T)
        '''
        MMT = M_k.dot(M_k.T) 
        if np.linalg.det(MMT) != 0:
            print(f'Step 3')
            #create sparse matrix
            MMT_sparse = sparse.csc_matrix(MMT)
            # if not MMT_sparse.sort_indices:
            #     MMT_sparse.sort_indices()

            lamda_k = Prob.lamda.copy() #copy by value
            factor = cholesky(MMT_sparse)
            lamda_k_W = factor(-d_k)

            print("W : \n ", Prob.W)
            print("M_k : \n ", M_k)
            lamda_k[Prob.W] = lamda_k_W
            print(f'lambda_k : \n {lamda_k}')

            if np.all(lamda_k >= -Prob.eps_d):
                print(f'Step 5')
                mu_k_W_hat = M_k_hat.dot(M_k.T.dot(lamda_k_W)) + d_k_hat
                mu_k[Prob.W_hat] = mu_k_W_hat
                Prob.lamda = lamda_k

                if np.all(mu_k > -Prob.eps_p):
                    print("Converged!")
                    break
                else:
                    min_mu = np.inf
                    j = None
                    for i_w_hat in Prob.W_hat:
                        if mu_k[i_w_hat]<min_mu:
                            min_mu =  mu_k[i_w_hat]
                            j = i_w_hat
                
                    # j = Prob.W_hat[np.argmin(mu_k_W_hat)]
                    Prob.W = np.hstack((Prob.W, j)) #check output
                    Prob.W_hat = np.setdiff1d(np.arange(Prob.m),Prob.W)
                    print(f'Step 7: New W \n {Prob.W}')

            else:
                print('Step 9')
                p_k = lamda_k - Prob.lamda
                #index included in working set where lamda<0
                B = [] 
                for idx in Prob.W:
                    if lamda_k[idx][0]< Prob.eps_d:
                        B.append(idx)

                B = np.array(B)
                print("B: ", B)                

                print('Step 14')
                lamda_next, W_next = fix_component(Prob.lamda, Prob.W, B, p_k)
                print("lamda_next: ", lamda_next)
                Prob.lamda = lamda_next
                Prob.W = W_next

        else:
            print(f'Step 12:')
            U, s, V = np.linalg.svd(MMT)
            # extract the nullspace from the decomposition
            nullspace = V[np.argwhere(s < 1e-10).flatten()]
            print(nullspace)
            p_k = None

            for col in nullspace.T:
                # print(col)
                p_k = np.zeros(Prob.m)
                # print(Prob.W)
                p_k[Prob.W] = col
                if col.dot(p_k) < Prob.eps_z:
                    break
            
            print(f'Step 13')
            B = []
            for idx in Prob.W:
                if p_k[idx] < Prob.eps_z:
                    B.append(idx)
            B = np.array(B)
            print(f'B : {B}')

            Prob.lamda, Prob.W = fix_component(Prob.lamda, Prob.W, B, p_k)

        k = k+1

    print(f'Step 16')
    x = -1 * Prob.Rinv.dot(M_k.T.dot(lamda_k_W) + Prob.v)
    
    return x, Prob.lamda, Prob.W

            
def fix_component(lamda_k, W_k, B, p_k):
    
    print(f'Step 18')
    min_val = np.inf
    j = 0

    for i in B:
        temp = -1 * lamda_k[i][0]/p_k[i][0]
        if temp<min_val:
            min_val = temp
            j = i
            
    print(f'Old W_k: \n {W_k} \n j = {j}')
    #removes j from W_k
    W_new = np.setdiff1d(W_k,j) 
    print("New W_k", W_new)
    lamda_new = lamda_k - (lamda_k[j]/p_k[j]) * p_k

    return lamda_new, W_new



if __name__ == "__main__":

    H = np.array([[65.0, -22, -16],
              [-22.0, 14, 7],
              [-16, 7, 5]])

    f = np.expand_dims(np.array([3.0, 2.0, 3.0]),axis=1)
    # print(f)
    
    A = np.array([[1.0, 2.0, 1.0],
              [2.0, 0.0, 1.0],
              [-1.0, 2.0, -1.0]])

    b = np.expand_dims(np.array([3.0, 2.0, -2.0]), axis=1)

    # Instantiating Object
    prob = Active_set(H, f, A, b)
    
    # Forming Dual Objective
    prob.form_dual_objective()
    
    ans, lamda, W = solve(prob)
    print(f'Solution: \n {ans}')

File: active_set.cpp
#include <iostream>
#include <eigen3/Eigen/Dense>
#include <vector>
#include <iterator> // for std::ostream_iterator
#include <algorithm>

std::ostream& operator<<(std::ostream& os, const std::vector<int>& v)
{
    os << "[";
    for (int i = 0; i < v.size(); ++i) {
        os << v[i];
        if (i != v.size() - 1)
            os << ", ";
    }
    os << "]\n";
    return os;
}
  

std::vector<int> arange(double low, double high)
{
    std::vector<int> W_hat;
    for(int i = low; i < high; i++){
        W_hat.push_back(i);
    }
    return W_hat;
}

class Active_set{

    public:
        Eigen::MatrixXf H; Eigen::MatrixXf A;
        Eigen::MatrixXf f; Eigen::MatrixXf b;
        
        // Dual Formulation of the Problem
        std::vector<int> W; std::vector<int> W_hat;
        Eigen::VectorXf lambda; Eigen::MatrixXf R;

        Eigen::MatrixXf M; Eigen::MatrixXf mu;
        Eigen::MatrixXf d;
        Eigen::MatrixXf v;

        //Primal Variables
        Eigen::MatrixXf x; 


        //Shapes
        int H_n, H_m;
        int A_n, A_m;
        int M_n, M_m;
        

    Active_set(Eigen::MatrixXf H, Eigen::MatrixXf A, Eigen::MatrixXf f, Eigen::MatrixXf b){
        
        this->H = H; this->A = A;
        this->b = b; this->f = f;

        Eigen::LLT<Eigen::MatrixXf> llt(this->H);
        Eigen::MatrixXf L = llt.matrixL();
        this->R = L.transpose();

        this->M = this->A * this->R.inverse();
        Eigen::MatrixXf Rt = this->R.transpose(); 
        this->v = Rt.inverse() * this->f;
        this->d = this->b + this->M * this->v;

        this->H_n = H.rows(); this->M_n = M.rows();
        this->H_m = H.cols(); this->M_m = M.cols();
        this->A_n = A.rows(); this->A_m = A.cols();

    }

    Eigen::MatrixXf Splice_Index(std::vector<int> &W, Eigen::MatrixXf &M){
        
        Eigen::MatrixXf M_w(W.size(), this->M_m);
        std::cout<<M<<std::endl;
        for(int i =0; i<W.size(); i++){
            // std::cout<<this->M.row(W[i])<<std::endl;
            M_w.row(i) = this->M.row(W[i]);
            // std::cout<<M_w.row(i)<<std::endl;
        }

        return M_w;

    }

    Eigen::MatrixXf replace_index(Eigen::MatrixXf lamda_k, Eigen::MatrixXf lamda_w_k, std::vector<int> W){
    
        int idx = 0;
        for(auto i: W){
            lamda_k.row(i) = lamda_w_k.row(idx);
            idx++;
        }

        return lamda_k;

    }

    bool all_ge_zero(Eigen::MatrixXf lamda){
    
        bool all_positive = (lamda.array() > 0).all();
        return all_positive;
    
    }

    int argmin(Eigen::MatrixXf mu, std::vector<int> W_hat){

        float min = 123123123;
        int j = -1;

        for(int i=0; i<W_hat.size(); i++){
            if(mu[W_hat[i]]<min){
                j = W_hat[i];
                min = mu[W_hat[i]];
            }
        }
        return j;
    }

    void remove(std::vector<int> &W, int k){
        auto it = std::find(W.begin(), W.end(), k);
        if (it != W.end()) { // if the value is found
            W.erase(std::remove(W.begin(), W.end(), k), W.end());
        }
    }

    std::vector<int> idx_le_zero(Eigen::MatrixXf lambda, std::vector<int> W){
    
        std::vector<int> B;
        for(int i=0; i<W.size(); i++){
            if(lambda[W[i]][0]<0)
                B.push_back(W[i]);
        }

        return B;
    }

    void fix_component(Eigen::MatrixXf lamda_k, std::vector<int> W, std::vector<int> B, Eigen::MatrixXf pk, Eigen::MatrixXf &lamda_next, std::vector<int> &W_next);

    void solve(){
    
        this->W.push_back(0);
        this->W.push_back(1);
        // std::cout<<this->M;
        this->W_hat = arange(2, this->M.rows());
        // std::cout<<this->W_hat<<std::endl;
        this->lambda = Eigen::VectorXf::Zero(this->M.rows());
        this->lambda[0] = 3;
        Eigen::MatrixXf mu_k;

        int k = 0;
        int j = -1;

        while(true){
            // std::cout<<"Working till here"<<std::endl; 
            Eigen::MatrixXf M_w = this->Splice_Index(this->W, this->M);
            Eigen::MatrixXf M_w_hat = this->Splice_Index(this->W_hat, this->M);

            Eigen::MatrixXf d_w = this->Splice_Index(this->W, this->d);
            Eigen::MatrixXf d_w_hat = this->Splice_Index(this->W_hat, this->d);
            
            Eigen::MatrixXf MMt = M_w * M_w.transpose();
            Eigen::MatrixXf lamda_k = this->lambda;

            if(MMt.determinant()!=0){
                // Non-Singular
                Eigen::MatrixXf lamda_w_k = MMt.inverse()*(-1*d_w);
                lamda_k = replace_index(lamda_k, lamda_w_k, this->W);
                
                if(all_ge_zero(lamda_k)){
                    Eigen::MatrixXf mu_k_hat = M_w_hat * M_w.transpose() * lamda_w_k + d_w_hat;
                    mu_k = replace_index(mu_k, mu_k_hat, this->W_hat);
                    this->lambda = lamda_k;

                    if(all_ge_zero(mu_k))
                        break;
                    else{
                        j = argmin(mu_k, this->W_hat);
                        this->W.push_back(j);
                        remove(this->W_hat, j);
                    }
                }
                else{
                    Eigen::MatrixXf pk = lamda_k - this->lambda;
                    std::vector<int> B = idx_le_zero(lamda_k, this->W);
                    Eigen::MatrixXf lamda_next;
                    std::vector<int> W_next;
                    fix_component(lamda_k, this->W, B, pk);
                    this->lambda = lamda_k;
                }
            }
            else{
                // Singular
                Eigen::MatrixXf mat = M_w * M_w.transpose();
                Eigen::FullPivHouseholderQR<Eigen::MatrixXf> qr(mat);
                Eigen::MatrixXf Q = qr.matrixQ();

                Eigen::MatrixXf null_space;
                if (qr.rank() < mat.cols()) {
                    null_space = Q.rightCols(mat.cols() - qr.rank());
                } else {
                    null_space = Eigen::MatrixXf::Zero(mat.rows(), 0);
                }

                Eigen::MatrixXf pk;

                for(int i=0; i<null_space.cols(); i++){
                    Eigen::MatrixXf col = null_space.col(i);
                    Eigen::MatrixXf pk_temp = replace_index(pk, col, this->W);
                    // CHECK SIZES
                    float dot = pk_temp.dot(this->d);
                    if(dot<0){
                        pk = pk_temp;  
                    }
                }

                std::vector<int> B = idx_le_zero(pk, this->W);
                fix_component(lamda_k, this->W, B, pk); 
                this->lambda = lamda_k;

            }

            k++;
        }

        Eigen::MatrixXf X = -1 * this->R.inverse() * (Splice_Index(this->W, this->M).transpose() * Splice_Index(this->W, this->lambda) + this->v);
        this->x = X;
        
    };

    void fix_component(Eigen::MatrixXf lamda_k, std::vector<int> &W, std::vector<int> B, Eigen::MatrixXf pk){
        
        int j = -1;
        float min = 123123123;
        for(int i = 0; i<B.size(); i++){
            if(-1*lamda_k[B[i]]/pk[B[i]] < min){
                min = -1*lamda_k[B[i]]/pk[B[i]];
                j = B[i];
            }
        }
        remove(W, j);
        lamda_k = lamda_k - (lamda_k[j]/pk[j])*pk;
    }
};

int main(){

    Eigen::MatrixXf H {
        {65.0, -22, 16},
        {-22,   14,  7},
        {-16,   7,   5}
    };

    Eigen::MatrixXf A {
        {1.0, 2.0, 1.0},
        {2.0, 0.0, 1.0},
        {-1.0, 2.0, -1.0}
    };

    Eigen::MatrixXf f{
                    {3.0}, 
                    {2.0}, 
                    {3.0}
    };

    Eigen::MatrixXf b{
                    {3.0}, 
                    {2.0}, 
                    {-2.0}
    };

    Active_set Prob(H, A, f, b);

    Prob.solve();
    std::cout<<Prob.x<<std::endl;

    return 0;
}
